subsys/net/l2_wifi_if_conn/l2_wifi_conn.c

```diff
@@ -0,0 +1,181 @@
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_event.h>
+#include <net/wifi_mgmt_ext.h>
+#include <net/l2_wifi_connect.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+#include <zephyr/src/supp_events.h>
+#include "conn_mgr_private.h"
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_CONNECT_RESULT | \
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+#define WPA_SUPP_EVENTS (NET_EVENT_WPA_SUPP_READY)
+
+static struct net_mgmt_event_callback net_l2_mgmt_cb;
+static struct net_mgmt_event_callback net_wpa_supp_cb;
+int connection_status;
+
+int net_l2_wifi_connect(struct conn_mgr_conn_binding *const binding);
+int net_l2_wifi_disconnect(struct conn_mgr_conn_binding *const binding);
+
+K_SEM_DEFINE(wpa_supp_ready_sem, 0, 1);
+
+/* Delayable work used to handle Wi-Fi connection timeouts. */
+static void wifi_conn_work_handler(struct k_work *work);
+K_THREAD_STACK_DEFINE(l2_wifi_wq_stack_area, CONFIG_L2_WIFI_CONN_WQ_STACK_SIZE);
+struct k_work_q wifi_conn_wq;
+static K_WORK_DELAYABLE_DEFINE(wifi_conn_work, wifi_conn_work_handler);
+
+/* Store a local reference to the connection binding */
+static struct net_if *wifi_iface;
+
+static void wifi_conn_work_handler(struct k_work *work)
+{
+	if (connection_status != WIFI_STATE_COMPLETED) {
+		net_l2_wifi_disconnect(NULL);
+	}
+}
+
+static int wifi_conn_timeout_schedule(void)
+{
+	int rc;
+
+	rc = net_mgmt(NET_REQUEST_WIFI_CONNECT_STORED, wifi_iface, NULL, 0);
+
+	if (rc) {
+		LOG_ERR("net management connect_stored request failed: %d\n", rc);
+		return -ENOEXEC;
+	}
+
+	if (conn_mgr_if_get_timeout(wifi_iface) > CONN_MGR_IF_NO_TIMEOUT) {
+		rc = k_work_schedule_for_queue(&wifi_conn_wq, &wifi_conn_work,
+					       K_SECONDS(conn_mgr_if_get_timeout(wifi_iface)));
+	}
+
+	return rc;
+}
+
+static void net_l2_wifi_connect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (status->status) {
+		connection_status = WIFI_STATE_DISCONNECTED;
+	} else {
+		connection_status = WIFI_STATE_COMPLETED;
+	}
+	k_work_cancel_delayable(&wifi_conn_work);
+}
+
+static void net_l2_wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
+				    uint32_t mgmt_event, struct net_if *iface)
+{
+	if (iface != wifi_iface) {
+		return;
+	}
+
+	switch (mgmt_event) {
+	case NET_EVENT_WIFI_CONNECT_RESULT:
+		net_l2_wifi_connect_result(cb);
+		break;
+	case NET_EVENT_WIFI_DISCONNECT_RESULT:
+		if (conn_mgr_if_get_flag(wifi_iface, CONN_MGR_IF_PERSISTENT)) {
+			k_work_reschedule_for_queue(&wifi_conn_wq, &wifi_conn_work,
+							K_SECONDS(conn_mgr_if_get_timeout(iface)));
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+static void wpa_supp_event_handler(struct net_mgmt_event_callback *cb,
+				    uint32_t mgmt_event, struct net_if *iface)
+{
+	switch (mgmt_event) {
+	case NET_EVENT_WPA_SUPP_READY:
+		k_sem_give(&wpa_supp_ready_sem);
+		break;
+	default:
+		break;
+	}
+}
+
+int net_l2_wifi_connect(struct conn_mgr_conn_binding *const binding)
```

Comment:
Is this function safe to call at boot? I've heard that there needs to be some delay between boot time and calling this so that the WPA supplicant can become ready

If that is the case, this needs to create a workqueue item that gets executed after the WPA supplicant becomes ready

I think there is a PR currently for supporting a notification of when the WPA Supplicant is ready -- until that PR is merged, this could be a simple delayable workqueue item with a delay of like one second.

In fact, I am pretty sure you could use `K_TIMEOUT_ABS_MS(1000)` to basically make all of those requests wait until 1 second after boot, but then execute instantly as long as it has been more than a second since boot

Copying @krish2718 on this just to keep you in the loop

Reply:
The PR for supporting WPA supplicant notifications has been merged. We have registered l2_wifi_conn module to listen on supplicant events. In init call, we are waiting on supplicant being ready. This handles the concern you raised about the connect call at bootup.